src/java.net.http/share/classes/jdk/internal/net/http/common/TimeSource.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.net.http.common;
+
+import java.time.Instant;
+import java.time.InstantSource;
+
+/**
+ * An {@link InstantSource} based on {@link System#nanoTime()} for the
+ * purpose of handling timeouts. This time source is intentionally not
+ * based on the system clock, in order to not be sensitive to clock skews
+ * caused by changes to the wall clock. Consequently, callers should use
+ * instants returned by this time source solely for the purpose of
+ * comparing them with other instants returned by this same time source.
```

Comment:
Do not use `Instant` as the value from this time source. It is not comparable with the real Instants and would be misleading lead to bugs.

Reply:
Wrapping the Instant is safe/fine.
However, it could quite a bit simpler (it seems) to just use the raw `long` nanoTime values.
Though milliseconds or seconds would be sufficient for the timeouts.
The complexity of the volatile read and needing to reset the nanoSource could be avoided.

Reply:
Working with Instant (now Deadline) and  Duration is much safer and easier than working with longs representing nano time, especially when you have a sorted list of deadlines.